validation: have LoadBlockIndex account for snapshot use

jamesob opened the pull request:
This is part of the [assumeutxo project](https://github.com/bitcoin/bitcoin/projects/11) (parent PR: #15606) 

---

Currently, `BlockManager::LoadBlockIndex` adds all blocks that have downloaded transactions to the active chain state's `setBlockIndexCandidates` set, ignoring the background chain state.

This PR changes ChainstateManager::LoadBlockIndex to update `setBlockIndexCandidates` in the background chain, not just the active chain. In the active chain, the same blocks are added as before. In the background chain, only blocks that have actually been validated, not blocks marked assumed-valid are added so the background chain will continue to download and validate assumed-valid blocks.

jonatack reviewed (commented):
After debug building rebased to master and an initial review, the code builds cleanly and makes sense, but I need to look more deeply and haven't yet checked if the change is covered by tests, so work-in-progress ACK 51e614fdbc6a23ade324e9bd01228bf97df9bc5d. A few minor comments, feel free to ignore.

ryanofsky reviewed (approved):
Code review ACK 51e614fdbc6a23ade324e9bd01228bf97df9bc5d. I left a lot of suggestions and some questions, but this change seems like a step in right direction and it shouldn't change behavior in any noticeable way when there's no snapshot loaded.

ryanofsky commented:
Forgot to post this earlier. I wrote up a description of this PR when I started reviewing it that may be useful to others (or helpful to me if I got something wrong):

Currently, `BlockManager::LoadBlockIndex` adds all blocks that have downloaded transactions to the _active_ chain state's `setBlockIndexCandidates` set, ignoring the _background_ chain state.

This PR changes `ChainstateManager::LoadBlockIndex` to:

- Update `setBlockIndexCandidates` in the background chain, not just the active chain. In the active chain, all blocks are added as before. In the background chain, only blocks below the snapshot height are added so the background chain will continue to download and validate them.
- Sets `pindex->nChainTx` of the snapshot block to the value from snapshot metadata instead of leaving it 0.

ariard reviewed (commented):
Concept ACK, as highlighted by other reviewers would be good to have unit tests to assert correctness of the changes.

jamesob committed:
validation: don't modify genesis during snapshot load

Avoid modifying the genesis block index entry during snapshot load. This
is because, in a future change that fixes LoadBlockIndex for UTXO
snapshots, we detect block index entries that are reliant on
assumed-valid ancestors and treat them specially.

Since the genesis block doesn't have BLOCK_VALID_SCRIPTS, it would be
erroneously marked BLOCK_ASSUMED_VALID during snapshot load if we didn't
skip it here. This would cause a "setBlockIndexCandidates() empty"
assertion to be tripped since all block index entries would be marked
assume-valid due to genesis, which is never re-validated.

There's probably no good reason to modify the genesis block index entry
during snapshot load anyway...

jamesob committed:
validation: have LoadBlockIndex account for snapshot use

Ensure that blocks past the snapshot base block (i.e. the end of the
assumed-valid region of the chain) are not included in
setBlockIndexCandidates for the background validation chainstate. These
blocks, while fully validated and lacking the BLOCK_ASSUMED_VALID flag,
*rely* on blocks which are assumed-valid, and so shouldn't be added to
the IBD chainstate.

Co-authored-by: Russ Yanofsky <russ@yanofsky.org>

jamesob commented on src/validation.cpp:
```diff
@@ -3658,7 +3691,18 @@ bool BlockManager::LoadBlockIndex(
                  (pindex->HaveTxsDownloaded() || pindex->pprev == nullptr))) {
-            block_index_candidates.insert(pindex);
+            for (CChainState* chainstate : chainman.GetAll()) {
+                // If we're on an block index entry that is passed the end of the
```
Fixed

ryanofsky reviewed (approved):
Code review ACK dbca47ece6b9a3a5e02969270bfab340fce286c0. Much simpler now and looks all good to me!

In the PR description https://github.com/bitcoin/bitcoin/pull/23174#issue-1015261643:

> When reconnecting the block index on startup, we have to account for the fact that assumed-valid blocks (per the snapshot chainstate) do not have nTx values. We can't exclude blocks from inclusion in `setBlockIndexCandidates` on this basis or we will trip over an empty set in an assertion later in startup.

I'm actually not sure what this is referring to, and it seems like it might not be applicable anymore according to https://github.com/bitcoin/bitcoin/pull/23174#discussion_r737666011 if nTx values are set and the previous code is now reverted

> We also have to ensure that blocks past the snapshot base block (i.e. the end of the assumed-valid region of the chain) are not included in setBlockIndexCandidates for the background validation chainstate.

The way this is phrased makes it sound like this change avoids adding blocks, when actually it is adding more blocks. Would maybe try to make it clearer this is now adding blocks for the background chainstate. Feel free to steal PR description I posted earlier too https://github.com/bitcoin/bitcoin/pull/23174#issuecomment-947075033 if that is useful

jamesob commented on src/chain.cpp:
```diff
@@ -119,6 +119,15 @@ void CBlockIndex::BuildSkip()
         pskip = pprev->GetAncestor(GetSkipHeight(nHeight));
 }
 
+bool CBlockIndex::reliesOnAssumedValid() const {
```
Yep, will test this.

Also, this change is not mergeable as-is: it turns out invoking this for each CBlockIndex in the chain takes a really. long. time. making startup on mainnet appear as though it is hung. :cry:

jamesob replied:
Fixed - nothing to test anymore.

jamesob commented on src/validation.cpp:
```diff
@@ -4912,6 +4912,14 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
     for (int i = 0; i <= snapshot_chainstate.m_chain.Height(); ++i) {
         index = snapshot_chainstate.m_chain[i];
 
+        if (index->isGenesis()) {
```
Done.

jamesob commented on src/validation.cpp:
```diff
@@ -4912,6 +4912,14 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
 
+        if (index->isGenesis()) {
+            // Don't make any modifications to the genesis block.
+            // This is especially important because we don't want to erroneously
```
Done.

jamesob commented on src/chain.h:
```diff
@@ -312,6 +312,10 @@ class CBlockIndex
 
+    // Runtime: O(n) with chain length.
+    // Returns true if the block specified relies on ancestors which are assumed-valid.
+    bool reliesOnAssumedValid() const;
```
Removed this method altogether, so done.

jamesob commented on src/validation.cpp:
```diff
@@ -3649,7 +3653,18 @@ bool BlockManager::LoadBlockIndex(
+                    // When blockindex doesn't rely on assumedvalid blocks, add it to all chainstates.
+                    chainstate->setBlockIndexCandidates.insert(pindex);
+                }
+            }
```
Addressed everything here aside from the unittest, which I will follow up with.

jamesob commented:
Man this changeset keeps getting smaller and smaller. Thanks @ryanofsky!

I've reverted back to doing a stateful approach for assumed-valid detection in LoadBlockIndex since the more holistic approach of scanning ancestors ended up being prohibitively slow when loading the main chain. This has allowed me to drop a few more changes from the original patch.

The only remaining item left is to write unittests for the contents of `setBlockIndexCandidates` after calling LoadBlockIndex, which I'll be working on shortly.

ryanofsky commented on src/validation.cpp:
```diff
@@ -3649,7 +3661,23 @@ bool BlockManager::LoadBlockIndex(
+            // the valid chain.
+            for (CChainState* chainstate : chainman.GetAll()) {
+                if (!seen_assumed_valid || chainstate->reliesOnAssumedValid()) {
+                    chainstate->setBlockIndexCandidates.insert(pindex);
```
In commit "validation: have LoadBlockIndex account for snapshot use" (2833bd0277ac87cd07e98b0a23b73eed97a2e513)

I still think it would be really nice to call this function from a unit test and make sure this set is filled correctly. The logic here is fragile, and testing this part of `LoadBlockIndex` could make it easier to test other parts as well.

sdaftuar replied:
Sorry for chiming in on what is now a very old PR, but I believe this block of code is incorrect. If we have a snapshot in place (and so some block index is marked IsAssumedValid), but we have downloaded blocks that haven't yet been activated (eg on the background-validation chainstate), this logic prevents us from adding those downloaded-but-not-yet verified blocks to `setBlockIndexCandidate` for the background chainstate. 

I ran into this while working on a download implementation for assumeutxo, and in some sequence of node restarts I got stuck in a state where my node had downloaded a bunch of blocks but never activated them, because they weren't in `setBlockIndexCandidates`.  I thought that restarting should fix the issue, but it didn't due to this block of code not doing the right thing...

Admittedly there is a lot of complicated interaction here but I think there's an invariant for `setBlockIndexCandidates` that is violated by this block of code, which is that if we have the data for a block and all its predecessors and it has more work than our tip, then it should be in `setBlockIndexCandidates`.  By excluding blocks that are marked IsAssumedValid() we violate this.

jamesob replied:
@sdaftuar:

> If we have a snapshot in place (and so some block index is marked IsAssumedValid), but we have downloaded blocks that haven't yet been activated (eg on the background-validation chainstate), this logic prevents us from adding those downloaded-but-not-yet verified blocks to setBlockIndexCandidate for the background chainstate

I'm having a hard time following how this could happen. If the bg chainstate has some downloaded-but-not-activated blocks, they are definitionally beneath the `first_assumed_valid_height` (since we don't attach blocks above that height to the bg chainstate). So any of those blocks should be added by this code to the bg chainstate's `setBlockIndexCandidates`. Have you seen an example where you're sure that a downloaded block on the bg chain is excluded by this code?

I don't doubt that you're seeing an issue though... I wonder if there might be a problem with how `m_best_header` is set below? I.e. that state isn't qualified per chainstate. That's something I'll investigate now.

I'd also be curious if anyone else has seen this issue when testing the assumeutxo tracking branch (https://github.com/bitcoin/bitcoin/pull/27596). Is there a chance some code you've added is causing this?

> By excluding blocks that are marked IsAssumedValid() we violate this.

Note that we're only excluding blocks marked assumedvalid (by way of the height check) for the bg chainstate, which still seems sensible to me. All blocks, assumedvalid or not, are added to the snapshot chainstate (or "regular" chain if no snapshot is in use).

sdaftuar replied:
> they are definitionally beneath the first_assumed_valid_height (since we don't attach blocks above that height to the bg chainstate)

I don't believe this is true; after we download a block and before we validate the block, the CBlockIndex entry for that block will have the AssumedValid bit set to true.  This means that if you were to download a bunch of blocks and then shutdown prior to those blocks being validated on the bg chainstate, that when you restart you'll be in this state.

sdaftuar replied:
The code here isn't comparing with the snapshot base height; it's just checking if a block index entry is below the `first_assumed_valid_height`, and only in that case does it get added to `setBlockIndexCandidates`.  [I don't believe there's any other logic elsewhere that will do what you are describing, of adding entries below the snapshot base height to `setBlockIndexCandidates`?]

sdaftuar replied:
More generally: I think that we don't need to worry at all about whether the assumed-valid bit is set when adding entries to `setBlockIndexCandidates`. We always validate a block when it gets connected, so the only invariant we need on `setBlockIndexCandidates` is whether they have enough work to be worth considering.  (We used to have a hard requirement that we always HAVE_DATA for such entries, but after pruning we had to relax that a bit, as there are edge-case scenarios where you might HAVE_DATA for such an entry and then prune it before you need it, so we handle that case by (re-)adding such entries to `mapBlocksUnlinked` when they occur. But I digress...)

jamesob replied:
> The code here isn't comparing with the snapshot base height; it's just checking if a block index entry is below the first_assumed_valid_height

Oops, total misread on my part. You're right; `first_assumed_valid_height` will be the block after the tip of the background chain.

ryanofsky replied:
The logic here does seem confusing.

Code comment says background candidate set needs to exclude BLOCK_ASSUMED_VALID blocks because otherwise ActivateBestChain "would add assumed-valid blocks to the chain". But that seems like exactly what it should do? (after validating them)

Commit message says background candidate set needs to exclude blocks after the snapshot height that don't have BLOCK_ASSUMED_VALID, but depend on blocks that do have BLOCK_ASSUMED_VALID. That idea does make sense, but does not seem to be what is implemented.

Taking another a step back, it seems like all of this would be simpler if background ActivateBestChain function just skipped calling FindMostWorkChain and tried to connect blocks up to the base of the snapshot.

jamesob committed:
test: add tests for LoadBlockIndex when using multiple chainstates

Incorporates feedback from Russ Yanofsky.

ryanofsky reviewed (approved):
Code review ACK 2833bd0277ac87cd07e98b0a23b73eed97a2e513. Changes since last review: Making PopulateAndValidateSnapshot loop from 1 instead of calling isGenesis. Also adding unit test. Making LoadBlockIndex use height comparison to determine assumed-valid blocks instead of traversing pprev.

re: https://github.com/bitcoin/bitcoin/pull/23174#issue-1015261643, https://github.com/bitcoin/bitcoin/pull/23174#issuecomment-947075033

I edited PR description to drop the part about nChainTx and stop referring to the snapshot height since implementation was changed to use first assumed-valid height, not last assumed-valid height. Previous description was:

> This is part of the [assumeutxo project](https://github.com/bitcoin/bitcoin/projects/11) (parent PR: #15606)
> 
> (PR description by ryanofsky)
> 
> Currently, BlockManager::LoadBlockIndex adds all blocks that have downloaded transactions to the active chain state's setBlockIndexCandidates set, ignoring the background chain state.
> 
> This PR changes ChainstateManager::LoadBlockIndex to:
> 
> Update `setBlockIndexCandidates` in the background chain, not just the active chain. In the active chain, all blocks are added as before. In the background chain, only blocks below the snapshot height are added so the background chain will continue to download and validate them. Sets pindex->nChainTx of the snapshot block to the value from snapshot metadata instead of leaving it 0.

jamesob commented:
@ryanofsky nice timing. I just pushed a unittest for `LoadBlockIndex()`/`setBlockIndexCandidates`.

jamesob commented on src/validation.cpp:
```diff
@@ -3622,17 +3622,29 @@ bool BlockManager::LoadBlockIndex(
-        // We can link the chain of blocks for which we've received transactions at some point.
+
+        if (!seen_assumed_valid && pindex->IsAssumedValid()) {
+            seen_assumed_valid = true;
```
Great suggestions, all integrated. Thank you!

ryanofsky reviewed (approved):
Code review ACK e35d167682bc527fa9d8734f1cc7558f18d95b09. Main change since last review was just adding test. And some suggestions from the last review were also applied.

This PR is now a lot simpler than it was at previous points, so it should be pretty easy for previous reviewers to reack, or new reviewers to take a look at.

ariard commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -309,4 +312,81 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_activate_snapshot, TestChain100Setup)
+    // The assumed-valid tolerant chain has all blocks as candidates.
+    BOOST_CHECK_EQUAL(cs2.setBlockIndexCandidates.count(validated_tip), 1);
+    BOOST_CHECK_EQUAL(cs2.setBlockIndexCandidates.count(assumed_tip), 1);
+    BOOST_CHECK_EQUAL(cs2.setBlockIndexCandidates.size(), num_indexes);
```
IIUC any fully validated block beyond the assumed-valid range of blocks shouldn't not be tolerated by the "fully-validated" chain, as they're relying on assumed-valid blocks ?

If so can you add a `BOOST_CHECK_EQUAL(cs1.setBlockIndexCandidates.count(m_chain.Height()), 0)` to verify this expected behavior holds ?

jamesob replied:
Did you mean `cs1.setBlockIndexCandidates.count(m_chain.Height())`? Because the set doesn't contain heights, it contains CBlockIndex*. Also which `m_chain` are you referring to? Because there's one for `c1` and `c2`. 

I think the check you want is already included in `BOOST_CHECK_EQUAL(cs1.setBlockIndexCandidates.count(assumed_tip), 0);`.

ariard commented on src/validation.cpp:
```diff
@@ -3622,17 +3622,30 @@ bool BlockManager::LoadBlockIndex(
+    int first_assumed_valid_height = std::numeric_limits<int>::max();
+    for (const auto& [height, block] : vSortedByHeight) {
+        if (block->IsAssumedValid()) {
+            first_assumed_valid_height = height;
```
nit: Maybe add a `assert(IsSnapshotActive()` as we shouldn't have assumed-valid blocks in the absence of a snapshot loaded ?

jamesob replied:
Part of @ryanofsky's overarching (and good) feedback a few months earlier was to avoid leaking knowledge of the idea of "UTXO snapshots" in places where it doesn't need to be. Instead we should rely on more fundamental state (like the presence of assumed-valid block index entries). I think referencing snapshots here would be going the other direction.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/23174/commits/b4919698ff3cc7183ae9ada7026d7300720f3fcc#r748949660

> nit: Maybe add a `assert(IsSnapshotActive()` as we shouldn't have assumed-valid blocks in the absence of a snapshot loaded ?

I guess like James said I'd be a little disinclined to add an assert checking for assumeutxo snapshots in validation code that's dealing more abstractly with chainstates. But I could also see the assert being useful as a sanity check, so no strong opinion either way.

Another option would be to write the assert a more generically with something like:

```c++

// assert that the assumed valid blocks come from some assumed-valid chainstate.
assert(any_of(chainstates, [](auto& chainstate) { return chainstate->reliesOnAssumedValid(); });

// assert that the assumed valid blocks will be validated by some not assumed-valid chainstate
assert(any_of(chainstates, [](auto& chainstate) { return !chainstate->reliesOnAssumedValid(); });
```

jamesob replied:
Good idea! Added.

ariard commented on src/validation.cpp:
```diff
@@ -3649,7 +3662,28 @@ bool BlockManager::LoadBlockIndex(
+            // Fill each chainstate's block candidate set. Add all blocks as
+            // candidates if the chainstate is allowed to rely on assumed-valid
+            // blocks. Otherwise avoid adding assumed-valid blocks so the
+            // chainstate will download and validate all earlier blocks.
```
I think the comment can be slightly modified to mention the descendants of assumed-valid blocks, like in the commit description.

"Otherwise avoid adding assumed-valid blocks and their fully-validated descendants therefore inducing the chainstate to download and validate all blocks earlier than the first assumed-valid block".

jamesob replied:
Not sure I follow your suggested comment, since fully-validated blocks should be added to both chainstates, but I've updated the comment to be try and be clearer.

ariard reviewed (commented):
The newly patchset is far simpler, just left few minor comments to confirm my understanding of the changes.

ryanofsky reviewed (approved):
Code review ACK bc3288b0ba1d949f0004513e250bb21ddfb5b389. Only change since previous review was fixing up a code comment to be more clear about which chainstate it was referring to

jamesob commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -312,4 +312,81 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_activate_snapshot, TestChain100Setup)
+    CTxMemPool& mempool = *m_node.mempool;
+    CChainState& cs1 = chainman.ActiveChainstate();
+
+    int num_indexes{0};
+    int num_assumed_valid{0};
+    int expected_assumed_valid{20};
+    int last_assumed_valid_idx{40};
+    int assumed_valid_start_idx = last_assumed_valid_idx - expected_assumed_valid;
```
Fixed, thanks

ryanofsky reviewed (commented):
Code review ACK 7a576ba9d24308e36dcde2b48b34c0676de98daa. Only change since previous review is adding new asserts and some tweaks in the test

fjahr reviewed (commented):
I reviewed the code and it looks very good to me with one potential optimization I suggested below.

I also did some mutation testing on the changes and discovered that the test in the first commit doesn't seem to cover the change. Unless I made some error here myself, the test should probably be fixed before merging.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -4909,7 +4952,14 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+    // Don't make any modifications to the genesis block.
+    // This is especially important because we don't want to erroneously
+    // apply BLOCK_ASSUMED_VALID to genesis, which would happen if we didn't skip
+    // it here (since it apparently isn't BLOCK_VALID_SCRIPTS).
```
I don't understand this comment. If genesis isn't `BLOCK_VALID_SCRIPTS`, then `BLOCK_ASSUMED_VALID` wouldn't be applied, so the workaround isn't needed??

jamesob replied:
No, *because* genesis isn't `BLOCK_VALID_SCRIPTS` we would apply `BLOCK_ASSUMED_VALID` unless we specifically skip it. The workaround is still desirable and necessary, I just bungled the unittest (see above).

MarcoFalke replied:
Yes, and the code below is:

```cpp
        if (!index->IsValid(BLOCK_VALID_SCRIPTS)) {
            // This flag will be removed once the block is fully validated by a
            // background chainstate.
            index->nStatus |= BLOCK_ASSUMED_VALID;
        }
```

So how could it be executed if `BLOCK_VALID_SCRIPTS` isn't set?

jamesob replied:
It is only executed when `BLOCK_VALID_SCRIPTS` *is* set - are you by chance missing the `!` in the conditional?

MarcoFalke replied:
> Edit: changed an "is" to "isn't" - the double negative got me!

Did you push the change or am I missing it?

MarcoFalke replied:
Nvm, apparently I don't understand how negation works.

jamesob replied:
This is an argument for including Python's `not` operator in C++21. ;)

ryanofsky replied:
> This is an argument for including Python's `not` operator in C++21. ;)

It's actually already there, no need to wait https://en.cppreference.com/w/cpp/language/operator_alternative

MarcoFalke replied:
Pretty sure that `not` may fail on some compilers (on windows?).

In general I think if there is a way to avoid a negation, it might be preferable. In this case saying "(since it apparently is lacking BLOCK_VALID_SCRIPTS)" might work?

jamesob commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -184,6 +184,9 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_activate_snapshot, TestChain100Setup)
     BOOST_CHECK(!chainman.SnapshotBlockhash());
 
+    // Ensure that the genesis block was not marked assumed-valid.
+    BOOST_CHECK(!chainman.ActiveChain().Genesis()->IsAssumedValid());
```
Oh yep, you're very right - we haven't actually yet activated the snapshot at this point in the test. Good find, thanks.

jamesob commented on src/validation.cpp:
```diff
@@ -3622,17 +3622,41 @@ bool BlockManager::LoadBlockIndex(
+
+            // assert that the assumed valid blocks come from some assumed-valid chainstate.
+            assert(any_of(chainstates.cbegin(), chainstates.cend(),
+                [](auto chainstate) { return chainstate->reliesOnAssumedValid(); }));
```
I don't understand your comment: these two asserts are only performed once if there is an assumed-valid block detected, and then the loop is broken out of (line 3642). We don't want to do the asserts if there is no assumed-valid block, so I don't see how or why we'd do them outside of the loop.

fjahr replied:
Unless I have a misunderstanding here, what I meant: We only have blocks that are assumed valid if we have a chainstate that relies on assume valid, those two will always go together. We will probably have very few chain states and we have a lot of blocks. So it seems to me that it would be more efficient to first loop the chainstates and then the blocks.

ryanofsky reviewed (approved):
Code review ACK 709a09c9b6a902d5ed0e5ad6ac421b17c8921e6e. Only change since last review is test change moving genesis not assume-valid check after the successful CreateAndActivateUTXOSnapshot call instead of before, so it is actually meaningful.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -3649,7 +3673,28 @@ bool BlockManager::LoadBlockIndex(
+            // beginning before the first assumed-valid block) might not get
+            // added to the the background chainstate, but this is ok, because
+            // they will still be attached to the active chainstate if they
+            // actually contain more work.
```
in commit 36e2bc4275fdb0d72390215fcea4d499f8049701:

I don't understand why it is safe to not attach them. If the assumed valid block is sufficiently old, it may cost little POW to create an alternative fork up to that height, and confusing the background chainstate that it can't connect to the active chainstate?

What is the reason for the if-condition in the first place? What would happen if both candidate sets are fully populated?

jamesob replied:
If both candidate sets are fully populated, when ActivateBestChain is called on the background validation chainstate, [`FindMostWorkChain()`](https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L2410-L2416) will return the assumed-valid block index entry as installed by snapshot activation, and it won't think it has to enter IBD to download and fully validate the chain up to the snapshot base block. I'll include this in the commentary above since people (myself included) seem to constantly forget why we need this conditional.

jamesob replied:
> If the assumed valid block is sufficiently old, it may cost little POW to create an alternative fork up to that height, and confusing the background chainstate that it can't connect to the active chainstate?

I'm confused by this because the background chainstate shouldn't be considering forks anyway; its purpose is to verify the assumed-valid blocks beneath the snapshot and it does not handle reorgs or forks.

MarcoFalke replied:
Ok, I guess I am missing context on how the download logic will be modified for assumevalid.

If the assumevalid chain is assumed to be always linear, then maybe a check that all assumed blocks are parents of the assumed tip block makes sense?

maflcko reviewed (commented):
Concept ACK 709a09c9b6a902d5ed0e5ad6ac421b17c8921e6e, did not look at the last commit (test) 🥇

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Concept ACK 709a09c9b6a902d5ed0e5ad6ac421b17c8921e6e, did not look at the last commit (test) 🥇
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUggmQv+IHs5tQZR4a2VV+VvpfYa4GEdcIbiiJjRJqReLcICNClFReu5hpKfkKm4
8VbU64iwXezNAeAzlJYUB7vWMKdGlHMMYX4V9fpJ6zwGkl38eL8Xq7kzHLUDH6iO
FfC32v0nywoz2FcnT5sFtAodnTl8kR01lQOCZSUvsvHnrUiYh6TGZt0Xlt1lz8Sh
2/Orx7mEYlmpR9lDPLrU6DJscUm12r5hE/PQnLC0iJ+nmc/kpKGV6S25fR08E0Id
JfPUnpzetTdiJGgjx3Zi9fau0ROmadXYPRJw8LSLm5cWCTxAt32oo4OQnoQeILy2
pmp76uQGwEnnD4vytiPg4J2ZBQUIzYz+X3f1DhWJhbDplfAcR1pKHO/f1itCqUhu
CUw5IbT1Wu4ocMepisosddMuoIuYSbb3Z3QXizXZHnD6E7WWKa1oNherTzX3IaFM
RW+rKuEZ0n1Wa0Sv5NV8OxxPQYLDwPXopHHz1K9Mz4/DJV6eOE9Q5+l1RtkpH9mr
0nul+aMW
=nvxK
-----END PGP SIGNATURE-----
```


</details>

Sjors commented on src/validation.h:
```diff
@@ -427,20 +427,16 @@ class BlockManager
     std::unique_ptr<CBlockTreeDB> m_block_tree_db GUARDED_BY(::cs_main);
 
-    bool LoadBlockIndexDB(std::set<CBlockIndex*, CBlockIndexWorkComparator>& setBlockIndexCandidates) EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
+    bool LoadBlockIndexDB(ChainstateManager& chainman) EXCLUSIVE_LOCKS_REQUIRED(::cs_main);
```
36e2bc4275fdb0d72390215fcea4d499f8049701: is it possible to spit this out as a refactor commit? IIUC `setBlockIndexCandidates` here is always just  `ActiveChainstate().setBlockIndexCandidates`.

jamesob replied:
It's possible, but I'm not going to do it because it creates too much intermediate churn (updating code that is changed anyway in the subsequent commit).

Sjors reviewed (commented):
utACK for the first commit 9be75f223577108bc573c2b04f2f8c67eefe88ef

One suggestion to make the second easier to review...

jamesob commented on src/validation.cpp:
```diff
@@ -3622,17 +3622,41 @@ bool BlockManager::LoadBlockIndex(
+            // assert that the assumed valid blocks will be validated by some not assumed-valid
+            // chainstate.
+            assert(any_of(chainstates.cbegin(), chainstates.cend(),
+                [](auto chainstate) { return !chainstate->reliesOnAssumedValid(); }));
```
> This condition should be independent of the loop and could be moved outside?

I don't see why you say that because (i) this condition should only be asserted if we encounter an assumed-valid block, which we need to loop through the index to find out, and (ii) the condition is only asserted once since we break immediately afterwards.

jamesob replied:
I've refactored this code and updated the commentary; hopefully you find it more intuitive now.

MarcoFalke replied:
With "condition" I mean "two chainstates exist".

Thus a loop of the form:

```py
for b in blocks:
  if b.assumed:
    assert len(chains())==2
```

Can be rewritten into:

```py
have_two_chains = len(chains())==2
for b in blocks:
  if b.assumed:
    assert have_two_chains
```

Obviously the performance should be the same here, but evaluating the condition outside the loop might be clearer. If would even allow to add additional checks. (Like asserting that *no* block is assumed if the number of chains is one?)

jamesob commented on src/validation.cpp:
```diff
@@ -3622,17 +3622,41 @@ bool BlockManager::LoadBlockIndex(
+            auto chainstates = chainman.GetAll();
+
+            // assert that the assumed valid blocks come from some assumed-valid chainstate.
+            assert(any_of(chainstates.cbegin(), chainstates.cend(),
```
Fixed.

jamesob commented on src/validation.cpp:
```diff
@@ -4909,7 +4909,14 @@ bool ChainstateManager::PopulateAndValidateSnapshot(
+    for (int i = AFTER_GENESIS_START; i <= snapshot_chainstate.m_chain.Height(); ++i) {
         index = snapshot_chainstate.m_chain[i];
 
         // Fake nTx so that LoadBlockIndex() loads assumed-valid CBlockIndex
```
Fixed, thanks.

Sjors commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -312,4 +312,81 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_activate_snapshot, TestChain100Setup)
+        auto index = cs1.m_chain[i];
+
+        if (i < last_assumed_valid_idx && i >= assumed_valid_start_idx) {
+            index->nStatus = BlockStatus::BLOCK_VALID_TREE | BlockStatus::BLOCK_ASSUMED_VALID;
```
Maybe add:
```cpp
    BOOST_CHECK(index->IsAssumedValid());
} else {
    BOOST_CHECK(index->!IsAssumedValid());
```

Sjors commented:
utACK 2283b9cd1ee0fbd1e8ebc61673b1fe7596199a24

Sjors reviewed (approved):
utACK 2283b9cd1ee0fbd1e8ebc61673b1fe7596199a24

MarcoFalke commented:
Concept ACK 2283b9cd1ee0fbd1e8ebc61673b1fe7596199a24 🤽

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

Concept ACK 2283b9cd1ee0fbd1e8ebc61673b1fe7596199a24 🤽
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjnlQwAmzp6vV9GqCu06YyivJ9zxvJ3Wsv+L8FxehFxMsVm3GTST2sM0bSfPL1M
6p9eBqXz3JgX6qwPaBsH8U5neWw75LkvpVMPMwmXQairRM3Odr9aGIyZJ5bOtCvv
OIMgrPsRz+bVMu1xGKcwCc3MM92LYRluGBAaGVq8qQTXSxZUfQWAoemqC+Ubhaiq
XXGJ9MPV02NOWD7ZK2qMTgNFsbPYM1zkVKX6r3kXVi0/jOBexZDrECJIcFvQZ/Ol
wqfv338xZnBYq3EnGkZVAiReSVJhpz7O+oMS/yKoHrtYhV07VRY15wxbZBXWaLgU
NlLqK5jZO49f2G5uaP2DgQ3CncMNbXEpG1lqJNxZy0bKUuuuio2GcgqJEWnSd7BN
f38nMQtQ80T9yUpfMW4/LWjj68XKROGbn9xLswIH654t+7DxHq63FzLq6jMFFto/
9K18qF1bEBwBpSX2sM9RxoTXT16W13sKxQnAjowcrA8HdAS5WXpVNGknyq01qEUE
bPxW9jNk
=FgfJ
-----END PGP SIGNATURE-----
```


</details>

laanwj commented:
It looks like this PR has had a good review cycle and there are no critical issues left. I think we should move toward a final round of ACKs and merge it. Smaller suggestions and nits that come up now can be addressed later.

MarcoFalke commented:
Ok, I'll open a follow-up, unless someone beats me to it.
